File: PERL/count_files.pl
#!/usr/bin/perl -w
=begin
U must have installed the following packages:
	sudo apt-get install build-essential
	sudo cpan DBI
	sudo apt-get install libpq-dev
	sudo cpan DBD::Pg module
=cut

use Benchmark qw(:all) ;
use DBI;
use warnings;
use strict;
use Data::Dumper;

#--Start-of-Benchmark--#
my $t0 = Benchmark->new;

#--Data-capture-from-DB--#
sub DataBase {

    #--Reading-DB-password--#
    my $password;
    open (PASS, "</etc/staffcop/config") or die "$!";
    while (<PASS>) {
        $password = "$1" if $_ =~ /(?<='PASSWORD': ')(.*)(?=',)/;
    }
    close (PASS);
    #-----------------------#

    my $driver = "Pg";
    my $database="staffcop";
    my $dsn = "DBI:$driver:dbname = $database; host = 127.0.0.1; port = 5432";
    my $userid = "staffcop";
    my $dbh = DBI->connect($dsn, $userid, $password, { RaiseError => 1 })
        or die $DBI::errstr;

    print "Opened database successfully!\n";

    my $queri = qq{SELECT data FROM agent_attachedfile};
    my $sth = $dbh->prepare($queri);
    my $rv = $sth->execute;
    my $data_db = $sth->fetchall_arrayref();

    $dbh->disconnect();

    return $data_db;
}

#--Slise-and-check-to-exist-raw-data-from-DB--#
sub Slice {

    my $get_ref = shift;
	my $external_storage; #Path a external storage
	my $full_list;
    my @fexist = ();

    foreach my $list_ref (@$get_ref) {
        foreach my $list (@$list_ref) {

            if ( $list =~ /^filedata/) {
                $full_list = "/var/lib/staffcop/upload/" . "$list";
            }
			elsif ( ! length $external_storage) {
				$external_storage = join("", split(/\/\d{4}_\d{2}_\d{2}.*/,,$list));
				return $external_storage;
			}
        }
    }
	print "Path to external storage: $external_storage";
}

#--Reading-and-Counting-files-from-../upload/filedata/..--#
sub OpenDirs {                                                                  
                                                                                
    my $upload_dirs = shift;                                                    
    my @folders;				#Variable for saving files                                  
    my @case;					#Variable for saving folders with files                     
                                                                                
    if ( -d $upload_dirs ) {                                                    
        opendir(local *DIR, $upload_dirs) || die "Error in opening dir $upload_dirs\n";
                                                                                
        foreach (readdir (DIR)) {                                               
            next if $_ eq "." or $_ eq ".." ;                                   
            my $sub_dir = "$upload_dirs" . "/" . "$_" ;                         
            push @folders, &OpenDirs($sub_dir);     #Push files in folder       
        }                                                                       
        closedir (DIR);                                                         
    }                                                                           
    else {                                                                      
        return $upload_dirs if -e $upload_dirs ;    #Retun a file                                   
    }                                                                           
                                                                                
    push @case, @folders;       #Push a folders with files                      
    return \@case;              #Return the results reference                   
}                                                                               
                                                                                
#Unwrap files into $output                                                      
sub Unwrap {                                                                    
                                                                                
    my $output = shift;                                                         
    my @path;                                                                   
    my $count = 0;                                                              
    foreach my $list (@$output) {                                               
        foreach my $list_ref (@$list) {                                         
            push @path ,$list_ref;                                              
            $count++;                                                           
        }                                                                       
                                                                                
    }                                                                           
    return \@path;                                                            
}                                                               

#--Search-Union-Difference-and-Crossing--#
sub DiVens {
	
	my $fexist_db = shift;	#Take a files reference from the DataBase
	my $phfiles = shift;	#Take a files reference from the Disk
	my %diff;
	
	#Difference $lesser_array\$great_array
	my %uniq_hash;
	print Dumper @$fexist_db;

	foreach my $list_ref (@$get_ref) {
        foreach my $list (@$list_ref) {

            if ( $list =~ /^filedata/) {
                $full_list = "/var/lib/staffcop/upload/" . "$list";
            }
			elsif ( ! length $external_storage) {
				$external_storage = join("", split(/\/\d{4}_\d{2}_\d{2}.*/,,$list));
		
	%uniq_hash = map { @{$_} => 1 if } @$fexist_db;				#Create a unique array

	foreach my $index (@$phfiles) {
		if ( ! $uniq_hash{$index} ) { $diff{$index} = 1; }		#Difference
	}

	my @delfiles = keys %diff;									#IT's the Difference for deleted
	
	#print Dumper \@delfiles;
	my $len = length(scalar(@delfiles));
	print "================"."="x$len."\n";
	print "||Difference: ", scalar(@delfiles),"||\n";
	print "================"."="x$len."\n";

	return \@delfiles;
}


#--Deleting-Difference-from-physical-storage--#
sub Deleting {

	my @upload = @{$_[0]};
	print  "\nNothing delete... \n" and goto SKIP if ! scalar(@upload);

	print "\nDeleting the files? (Yes/No): \n";
	my $choose = <>;

	if ( $choose =~ "[Yy]" ) {
		
		print join("\n", @upload);
		my $count = (unlink (@upload) or die "Having trouble deleting $upload[$_]: $!");
		system ("staffcop restart");
		print "\nDeleted is $count files\n";
	}
	else { print "Files is not deleting...\n"; }
}

##--Start-code-of-this--##

	exec("sudo $0") if ! $< == 0;				#Run script with "sudo"


	#--Pull-Files-From-DB--#
	my $ref_filebd = &DataBase();				#Pull all a files from DB
	my $ref_fexist_db = $ref_filebd; 			#Pull exist files
	my $ext_storage = &Slice($ref_filebd);		#

	#--Pull-Files-From-Physycal-Storage--##
	my $def_path = "/var/lib/staffcop/upload/filedata/by_date";  
	my $def_output = &OpenDirs($def_path);                          
	my $def_ref_phfiles = &Unwrap($def_output);                                                  
	my $full_ref_phfiles;

	#If files exist in external storage the functions is run
	if ( -d $ext_storage) {
		my $ext_output = &OpenDirs($ext_storage);
		my $ext_ref_phfiles = &Unwrap($ext_output);
		$full_ref_phfiles = [@$def_ref_phfiles, @$ext_ref_phfiles];
	}

	else { $full_ref_phfiles = [@$def_ref_phfiles]; }	#Finally links the physical files


	#--Print-Output--#
	printf "\n\nSummary existing files into Storage: %s\n", scalar(@$full_ref_phfiles);
	printf "Summary existing files from DataBases: %s\n", scalar(@$ref_fexist_db);
	printf "Difference: %s\n", ($#$full_ref_phfiles - $#$ref_fexist_db);

	#--Files-with-out-links--#
	my $ref_orf_files = &DiVens($ref_fexist_db,$full_ref_phfiles);

	#--Invoke-sub-Deleting--#
	&Deleting ($ref_orf_files);

##-------------------------##
SKIP:
#--Block-of-Benchmark--#
my $t1 = Benchmark->new;
my $td = timediff($t1, $t0);

print "\nTime execution of code: ", timestr($td), "\n";

File: PYTHON/count_files.py
#!/usr/share/staffcop/env18/bin/python2

import psycopg2
import os.path
import re
from datetime import datetime 
import time 
start_time = datetime.now()


full_path = '/var/lib/staffcop/upload/filedata/by_date/'
db_exists_files = set()

def_path='/var/lib/staffcop/upload/'
piece_path = ''
select = "SELECT data FROM agent_attachedfile"

def Get_pass():
    with open('/etc/staffcop/config', 'r') as f:
            strings = f.read()
            result = re.findall("(?<='PASSWORD': ')(.*)(?=',)", strings)
    return result[0]


class Data_Bases:

    def __init__(self, password):
        conn = psycopg2.connect("dbname=staffcop user=staffcop password={}".format(password))
        self.cur = conn.cursor()
        print "DataBase is open"

    def request(self, sql):
        self.cur.execute(sql)
        return self.cur.fetchall()

    def close_con(self):
        self.cur.close()
        self.conn.close()
        print "\nDataBase is Closed\n"

def Pull_files(path):
    array_files = set()
    for i in os.listdir(path):
        #print full_path + i
        if os.path.isdir("{}{}".format(path,i)):
            for j in os.listdir("{}{}".format(path, i)):
                #print "{}{}/{}".format(path, i, j)
                array_files.add("{}{}/{}".format(path, i, j))
        #elif  os.path.isfile(path + i):
    return array_files

def Deleting(supply):
    count = 0
    for path in supply:
        try:
            os.remove(path)
            print "{} deleting!".format(path)
            count += 1
        except Exception as e:
            count -= 1
            print e
    return count


password = Get_pass()
connection = Data_Bases(password)
raw_files = connection.request(select)
connection.close_con


## Unwrap raw data from DB ##
for i in raw_files :
    for j in i :
        if j.startswith(tuple('filedata')) :
                full_string = "{}{}".format(def_path, j)
        else :
            if len(piece_path) == 0:
                piece_path =  j.split('filedata')[0]
            full_string = "{}".format(j)
        db_exists_files.add(full_string)

print "Total Files From DB: {}".format(len(db_exists_files))


##Pull files from storage
def_storage_files = Pull_files(full_path)
size_def_storage = len(def_storage_files)

if os.path.isdir(piece_path):
	external_path = '{}filedata/by_date/'.format(piece_path)
	external_storage_files = Pull_files(external_path)
        for i in external_storage_files :
            def_storage_files.add(i)
	print "Total number Files into External Storage: {}".format(len(external_storage_files))



print "Total number Files into Default Storage: {}".format(size_def_storage)
print "Total Exist Files From Storage: {}\n".format(len(def_storage_files))

files_for_deleting = def_storage_files.difference(db_exists_files)
#files_for_deleting2 =  set_db_exists_files - set_storage_files

print "\nOrfaned Files Into Storage: {}".format(len(files_for_deleting))

choose = raw_input('Are you Want to delete files? (Yes/No) ')
if ("y" or "Y") in choose:
	print "DELETING {} Files.".format(Deleting(files_for_deleting))

print(datetime.now() - start_time)
